### PR TITLE
[ReuseIR] add basic alias analysis and fusion

### PR DESCRIPTION
**Pull request by SchrodingerZhu**

### CODE REVIEW

File: reuse-mlir/src/cxx/Passes/AcquireDestroyFusion.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+          acquireOps.push_back({acq, proj.getObject(), proj.getIndex()});
+  });
+  for (auto &destroy : destroyOps) {
+    auto ref = destroy.getFirst().getObject();
+    for (auto acq : acquireOps) {
+      if (/* destroy the container of an acquire operation */ aliasAnalysis
+                  .alias(ref, acq.projSource) == AliasResult::MustAlias &&
+          /* the acquire operation dominates the destroy */
+          domInfo.dominates(acq.acquire, destroy.getFirst()) &&
+          /* avoid repeated analysis */ !toErase.contains(acq.acquire) &&
```
Comment (SchrodingerZhu) [inline reuse-mlir/src/cxx/Passes/AcquireDestroyFusion.cpp:62]: @Lancern should I lift this `contains` test earlier?

File: reuse-mlir/src/cxx/Passes/AcquireDestroyFusion.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+          {destroy, SmallVector<int64_t>{destroy.getFusedIndices().begin(),
+                                         destroy.getFusedIndices().end()}});
+    else if (auto acq = dyn_cast<RcAcquireOp>(op))
+      if (auto load = dyn_cast<LoadOp>(acq.getRcPtr().getDefiningOp()))
+        if (auto proj = dyn_cast<ProjOp>(load.getObject().getDefiningOp()))
+          acquireOps.push_back({acq, proj.getObject(), proj.getIndex()});
+  });
+  for (auto &destroy : destroyOps) {
+    auto ref = destroy.getFirst().getObject();
+    for (auto acq : acquireOps) {
```
Comment (SchrodingerZhu) [inline reuse-mlir/src/cxx/Passes/AcquireDestroyFusion.cpp:57]: Another way is to walk the domtree. I don't think MLIR's dominfo provides that functionality.

File: reuse-mlir/src/cxx/Passes/AcquireDestroyFusion.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  // Collect operations to be considered by the pass.
+  DenseMap<DestroyOp, SmallVector<int64_t>> destroyOps;
+  struct AcquireTuple {
+    RcAcquireOp acquire;
+    mlir::Value projSource;
+    size_t projIndex;
+  };
+  SmallVector<AcquireTuple> acquireOps;
+  DenseSet<RcAcquireOp> toErase;
+  // TODO:
```
Comment (SchrodingerZhu) [inline reuse-mlir/src/cxx/Passes/AcquireDestroyFusion.cpp:44]: remove this line

**Comment by SchrodingerZhu** [comment]
Should be simplified to acquire release fusion. I will do it tomorrow.

File: reuse-mlir/src/cxx/Passes/AcquireDestroyFusion.cpp (hunk 1 above)
Comment (Lancern) [reply reuse-mlir/src/cxx/Passes/AcquireDestroyFusion.cpp:62]: Yes, lifting this `contains` seems more reasonable.

Comment (SchrodingerZhu) [reply reuse-mlir/src/cxx/Passes/AcquireDestroyFusion.cpp]: done

**Pull request merged**